### DISCUSSION TITLE
Why was SQLite removed?

### DISCUSSION THREAD

**Discussion by septatrix** [kimai/kimai#2977] [General]
I wanted to start using kimai2 on a cheap VPS using SQLite and noticed that support was removed in 1.14. Looking at the blog post and afterwards the issues it seems like most problems stem from cascading deletes though based on the [SQLite docs for foreign keys](https://www.sqlite.org/foreignkeys.html#fk_actions) they are long supported. Were there any other problems I have missed?

Due to the low power of the VPS it is not feasible for me to use a database server and I would also miss out on the simple backup process of only having to copy a single file when using SQLite. For now I sadly consider to either install the old version as I do not need any never features or switch to an alternative like traggo...

**Comment by kevinpapst** [comment]
Thanks for your feedback. The decision was explained in the blog post. The decision will not be reverted unless someone pays me good for the support.

I am probably not good in explaining myself in English, but you know what I care about besides Kimai: my time. It is extremely valuable and only available once. So even arguing about a decision that was already made is (for me) a waste of that precious gift that we all were given at birth. 

I am not going to spent more time in supporting a system that is used by a small percentage of the user base. If you are stuck with a bad hosting, I am sorry for you, but you should consider to choose a different software.

**Reply by septatrix** [reply]
In the blog you only stated that it was never supported for production purposes however you did not state which exact issues lead you to remove the support even for developing/testing. I understand that you value your time though I personally would be interested to fix this and invest a bit of my own time. Also I think that a big portion of your users have small scale deployments and use SQLite for simplicity.

According to the GitHub issues the main problem seems to be foreign key support. I found an upstream bug from doctrine about this (https://github.com/doctrine/dbal/issues/4288). Would you consider readding support if that issue were fixed?

**Reply by kevinpapst** [reply]
Thanks for you calm answer 😄 You are right, the blog does not contain all arguments - they where written in the documentation pages about SQLite, which were removed in the meantime.

I believe that you don't see the full picture, which is understandable if you never tried to support SQLIte and the user base in an application of the size of Kimai. 

These two points together are already a killer argument against:

- `In order to use foreign key constraints in SQLite, the library must be compiled with neither SQLITE_OMIT_FOREIGN_KEY or SQLITE_OMIT_TRIGGER defined.` - many (and I mean a HUGE amount of them) SQLite installations out there do NOT support foreign keys
- foreign key support for SQLite was activated in Kimai when available, but in MANY installations it didn't work and it caused the same support requests being posted over and over again, which sucked big times: user lost data and/or access to Kimai and I had to take care about support requests caused by vendor issues and users not reading docs (insert my argument about time here)

Then SQLite does not support the most basic DDL statements, which causes difficult and error prone migrations for simple add / drop column statements. Data can get lost during migrations and/or they take forever.

I have tried and it is definitely not worth the effort. See, I invest a HUGE amount of my lifetime into Kimai. If a user cannot pay 3 € a month for a small hosting with database, I am sorry but then this user is not the target group for the on-premise version of Kimai. 
There are other apps out there which are great as well and which will probably support SQLite. 

To wrap it up: SQLite will not be re-added to Kimai.

I understand your frustration and I understand every user that is pissed off by my decision. But as maintainer I have to decide for the future of this project and removing broken parts makes time available for other features. 

You can use the free version of Kimai https://www.kimai.cloud/ if you like it but don't have enough money for proper self-hosting.
